### PR TITLE
Dynamically configure verbose logging

### PR DESCRIPTION
**Pull request by kargakis**
Unify enabling verbose logging across the main prow
workhorses and add missing logging calls in Jenkins
and Slack clients.

Tested this with tide and it works like a charm.

/assign spxtr

### CODE REVIEW

File: prow/cmd/hook/main.go [hunk 1]
```diff
@@ -152,9 +151,11 @@ func main() {
 		logrus.WithError(err).Fatal("Error getting git client.")
 	}
 
+	logger := logrus.StandardLogger()
```
Comment (kargakis) [inline prow/cmd/hook/main.go]: I've left hook out for now because I didn't feel strong in adding a stanza for it in the config but eventually we may need to as other things come up.

File: prow/config.yaml [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
   report_template: '[Full PR test history](https://k8s-gubernator.appspot.com/pr/{{if eq .Spec.Refs.Org "kubernetes"}}{{if eq .Spec.Refs.Repo "kubernetes"}}{{else}}{{.Spec.Refs.Repo}}/{{end}}{{else}}{{.Spec.Refs.Org}}_{{.Spec.Refs.Repo}}/{{end}}{{with index .Spec.Refs.Pulls 0}}{{.Number}}{{end}}). [Your PR dashboard](https://k8s-gubernator.appspot.com/pr/{{with index .Spec.Refs.Pulls 0}}{{.Author}}{{end}}). Please help us cut down on flakes by [linking to](https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/{{.Spec.Refs.Org}}/{{.Spec.Refs.Repo}}/issues?q=is:issue+is:open) when you hit one in your PR.'
+  enable_verbose_log: false
```
Comment (spxtr) [inline prow/config.yaml]: I think it can just be called "verbose_log"

File: prow/config/config.go [hunk 3]
```diff
@@ -57,6 +57,8 @@ type Tide struct {
 	// These must be valid GitHub search queries. They should not overlap,
 	// which is to say two queries should never return the same PR.
 	Queries []string `json:"queries,omitempty"`
+	// EnableVerboseLog enables verbose logging for tide dynamically.
+	EnableVerboseLog bool `json:"enable_verbose_log,omitempty"`
```
Comment (spxtr) [inline prow/config/config.go]: Do we want one "verbose_log" config option for all of prow, or an individual option for each component?

Comment (kargakis) [reply prow/config/config.go]: Good question, I thought about this and don't feel strong in any way.

Single option:
- We don't bloat the config.
- Easy to use when you want to debug an event e2e.

Each component has its own:
- Finer-grained logging. We may want to run hook with high verbosity and jenkins-operator turned off.

I am willing to try out the single option now that I have layed out my thoughts on the pros of each side. Looking back at hook logging, it feels to me that high verbosity on it is not necessary by default.

Comment (kargakis) [reply prow/config/config.go]: Hrm, re hook it's not possible to force it update its client w/o hacking in it a bit more so it's left on by default for now.

File: prow/config.yaml (hunk 2 above)
Comment (kargakis) [reply prow/config.yaml]: fixed

**Comment by kargakis** [comment]
@spxtr any more comments here?

**Comment by spxtr** [comment]
I have mixed feelings about this one. Reaching in and setting the clients' `Logger` fields seems fragile. If, for instance, the tide controller decided to say `c.kc = c.kc.Namespace("some-other-ns")` then it would stop getting logging updates. This is probably not going to happen, but it reflects poorly on the design.

I haven't thought up a better way, though.

**Comment by kargakis** [comment]
It's not any different from today, is it? Rewritting the client is going to break any option that is set during initialization. I think this is the simplest and cleanest solution we can get right now for dynamic logging and I would really like to have it as an option when things go south.

Related discussion in https://github.com/kubernetes/kubernetes/issues/39758.

**Comment by spxtr** [comment]
I guess one alternative is to have everything use the standard logger instead of setting `c.Logger`. Then, when the `config.Agent` reloads the config it can set the default logger level.

I'm okay with this PR as well, though. Just not super excited about it.

**Comment by kargakis** [comment]
> I guess one alternative is to have everything use the standard logger instead of setting c.Logger. Then, when the config.Agent reloads the config it can set the default logger level.

We don't use different log levels as much today and I am more interested in disabling logging altogether, or more precisely enable logging when I want to debug something. Or we could change the client loggers to always print in debugging mode. Then, your suggestion makes sense. Thoughts?

**Comment by kargakis** [comment]
@spxtr updated as per your suggestion. Feels cleaner indeed.

File: prow/config/config.go [hunk 4]
```diff
@@ -53,6 +54,10 @@ type Config struct {
 	// Defaults to "default".
 	PodNamespace string `json:"pod_namespace,omitempty"`
 
+	// LogLevel enables dynamically updating the log level of the
+	// standard logger that is used by all prow components.
+	LogLevel string `json:"log_level,omitempty"`
```
Comment (spxtr) [inline prow/config/config.go]: Somehow indicate that it must be one of [these](https://github.com/sirupsen/logrus/blob/89742aefa4b206dcf400792f3bd35b542998eb3b/logrus.go#L36-L54). Maybe just explicitly list them here?

File: prow/config/config.go [hunk 5]
```diff
@@ -224,6 +229,15 @@ func parseConfig(c *Config) error {
 	if c.PodNamespace == "" {
 		c.PodNamespace = "default"
 	}
+
+	level := logrus.InfoLevel
+	if lvl, err := logrus.ParseLevel(c.LogLevel); err != nil {
+		logrus.Warningf("Cannot parse log level, will default to 'info': %v", err)
```
Comment (spxtr) [inline prow/config/config.go]: Propagate up the error rather than allowing it by, and special-case empty string to mean `logrus.InfoLevel`.

File: prow/cmd/jenkins-operator/main.go [hunk 6]
```diff
@@ -100,6 +100,11 @@ func main() {
 		ghc = github.NewClient(oauthSecret, *githubEndpoint)
 	}
 
+	logger := logrus.StandardLogger()
+	kc.Logger = logger.WithField("client", "kube")
+	jc.Logger = logger.WithField("client", "jenkins")
+	ghc.Logger = logger.WithField("client", "github")
```
Comment (spxtr) [inline prow/cmd/jenkins-operator/main.go]: It may be worth putting `gc.logger = logrus.StandardLogger().WithField("client", "github")` into `github.NewClient`. The `Logger` field doesn't really need to be public, and this helps make sure that things are consistent going forward.

I'd be okay with that happening in a separate PR.

File: prow/config/config.go (hunk 4 above)
Comment (kargakis) [reply prow/config/config.go]: done

File: prow/config/config.go (hunk 5 above)
Comment (kargakis) [reply prow/config/config.go]: done

File: prow/cmd/jenkins-operator/main.go (hunk 6 above)
Comment (kargakis) [reply prow/cmd/jenkins-operator/main.go]: Yeah, separate PR.

**Comment by spxtr** [comment]
Great stuff, thanks.

/lgtm
